### PR TITLE
Feature/dedupe imports

### PR DESCRIPTION
**Pull request by DovydasNavickas**
Solves:
#16, #17, #19

### CODE REVIEW

**Comment by DovydasNavickas** [comment]
@MartynasZilinskas, @GiedriusGrabauskas, please review the changes.
We have to check if there are breaking changes and should this be a major or a minor version.

File: src/bundler.ts [hunk 1]
```diff
@@ -20,34 +21,61 @@ export interface ImportData {
...
+    usedImports?: { [key: string]: number };
+    // Imports dictionary by file
+    importsByFile?: { [key: string]: BundleResult[] };
 }
 
 export class Bundler {
-    public static async Bundle(file: string, filesRegistry: Registry = {}): Promise<BundleResult> {
+    public static async BundleAll(
+        files: string[],
+        dedupeFiles: string[],
```
Comment (MartynasZilinskas) [inline src/bundler.ts]: Arguments breaking change.

Comment (MartynasZilinskas) [inline src/bundler.ts]: `dedupeFiles` -> `dedupeGlobs` maybe?

File: src/bundler.ts [hunk 2]
```diff
@@ -20,34 +21,61 @@ export interface ImportData {
...
+        fileRegistry: FileRegistry = {}
+    ): Promise<BundleResult[]> {
+        let resultsPromises = files.map(file => this.Bundle(file, dedupeFiles, fileRegistry));
+        return await Promise.all(resultsPromises);
+    }
+
+    public static async Bundle(file: string, dedupeGlobs: string[] = [], fileRegistry: FileRegistry = {}): Promise<BundleResult> {
```
Comment (MartynasZilinskas) [inline src/bundler.ts]: Another arguments breaking change.

File: src/bundler.ts [hunk 3]
```diff
@@ -81,52 +110,155 @@ export class Bundler {
...
+            filePath: filePath,
+            found: true
+        };
+
+        // if this is the root BundleResult
+        if (rootBundleResult == null) {
+            // Initialize usedImports array
+            bundleResult.usedImports = {};
+            bundleResult.importsByFile = {};
+        }
```
Comment (MartynasZilinskas) [inline src/bundler.ts]: Instead of saving into rootBundleResult, you can remove static methods and use class properties instead?

**Comment by DovydasNavickas** [comment]
I guess because we have breaking changes, I will update the `Bundler` not to be static. Because `usedImports` and `importsByFile` now has to reside on the `rootBundleResult`, which is not perfect at all.

File: src/arguments.ts [hunk 4]
```diff
@@ -9,6 +9,8 @@ for (let key in Contracts.Verbosity) {
     }
 }
 
+const dedupeGlobsKey = "dedupeGlobs";
```
Comment (GiedriusGrabauskas) [inline src/arguments.ts]: Uppercase

File: src/bundler.ts [hunk 5]
```diff
@@ -20,34 +21,61 @@ export interface ImportData {
...
 }
 
 export class Bundler {
-    public static async Bundle(file: string, filesRegistry: Registry = {}): Promise<BundleResult> {
+    public static async BundleAll(
+        files: string[],
+        dedupeFiles: string[],
+        fileRegistry: FileRegistry = {}
+    ): Promise<BundleResult[]> {
+        let resultsPromises = files.map(file => this.Bundle(file, dedupeFiles, fileRegistry));
```
Comment (GiedriusGrabauskas) [inline src/bundler.ts]: const

Comment (DovydasNavickas) [inline src/bundler.ts]: Done

File: src/arguments.ts (hunk 4 above)
Comment (DovydasNavickas) [inline src/arguments.ts]: Done

**Pull request merged**